src/skagent/algos/maliar.py

```diff
@@ -516,46 +607,184 @@ def get_bellman_equation_loss(
 
         Parameters
         ----------
+        vf : callable
+            Value function from value network
         df : callable
             Decision function from policy network
         input_grid : Grid
-            Grid containing current states and two independent shock realizations:
-            - {shock_sym}_0: period t shocks
-            - {shock_sym}_1: period t+1 shocks (independent of period t)
+            Grid containing states and two independent shock realizations
 
         Returns
         -------
         torch.Tensor
-            Bellman equation residual loss (squared)
+            Bellman residual loss (squared)
         """
+        # Use estimate_bellman_residual for computation
         given_vals = input_grid.to_dict()
 
-        # Extract current states and both shock realizations
+        # Extract current states and combined shocks for estimate_bellman_residual
         states_t = {sym: given_vals[sym] for sym in state_variables}
-        shocks = {f"{sym}_0": given_vals[f"{sym}_0"] for sym in shock_syms}
-        shocks.update({f"{sym}_1": given_vals[f"{sym}_1"] for sym in shock_syms})
+        shocks = given_vals  # Contains both _0 and _1 shock realizations
 
-        # Use helper function to estimate Bellman residual with combined shock object
+        # Compute Bellman residual using the existing function with actual value function
         bellman_residual = estimate_bellman_residual(
-            block,
-            discount_factor,
-            value_network,
-            df,
-            states_t,
-            shocks,
-            parameters,
-            agent,
+            block=block,
+            discount_factor=discount_factor,
+            vf=vf,
+            dr=df,
+            states_t=states_t,
+            shocks=shocks,
+            parameters=parameters,
+            agent=agent,
         )
 
         # Return squared residual as loss
         return bellman_residual**2
 
-    return bellman_equation_loss
+    return bellman_loss
+
+
+def bellman_training_loop(
+    block,
+    loss_function,
+    states_0_n: Grid,
+    parameters,
+    shock_copies=2,
+    max_iterations=5,
+    tolerance=1e-6,
+    random_seed=None,
+    simulation_steps=1,
+    epochs=250,
+):
+    """
+    Bellman training loop for joint policy and value network training.
+
+    This function performs joint training of both policy and value networks using
+    the full Bellman-residual minimization with all-in-one expectation operator.
+    For policy-only training with lifetime reward loss, use maliar_training_loop instead.
+
+    This implements joint optimization of θ ≡ (θ₁, θ₂) where:
+    - θ₁: Value function parameters V(·; θ₁)
+    - θ₂: Decision rule parameters φ(·; θ₂)
+
+    Parameters
+    ----------
+    block : DBlock
+        The model definition
+    loss_function : callable
+        A unified loss function with signature (value_function, decision_function, input_grid) -> loss.
```

Comment:
The difference in the loops seems to be about:
- the signature of the loss function
- the initialization of multiple networks.

I wonder if, in principle, it's possible to combine both the policy and value networks into a single network. Is that silly?

I'm asking because I recall that the MMW '21 algorithm is generic to the AiO operator used.